Add distributed data parallel benchmark tool

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35198 Add distributed data parallel benchmark tool**

The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

This benchmark was written by @pietern, and was originally submitted in #30660

Differential Revision: [D20591296](https://our.internmc.facebook.com/intern/diff/D20591296)

mrshenli committed:
Add distributed data parallel benchmark tool

The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

[ghstack-poisoned]

mrshenli committed:
Update on "Add distributed data parallel benchmark tool"


The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

This benchmark was written by @pietern, and was originally submitted in #30660

Differential Revision: [D20591296](https://our.internmc.facebook.com/intern/diff/D20591296)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM, would be useful to get this in. It looks like there are some CI failures currently.

Commit:
Update on "Add distributed data parallel benchmark tool"


The need for this tool was motivated by #28883. In the past, we have
done ad-hoc benchmarking, but it's time for something more structured.

It would be nice to add more model architectures so that we can get a
full picture of the performance impact of a code change simply by
running this suite a few times.

This benchmark was written by @pietern, and was originally submitted in #30660

Differential Revision: [D20591296](https://our.internmc.facebook.com/intern/diff/D20591296)

[ghstack-poisoned]
